promql/engine.go

```diff
@@ -1332,11 +1385,76 @@ func (ev *evaluator) eval(expr parser.Expr) (parser.Value, storage.Warnings) {
 			newEv.startTimestamp += newEv.interval
 		}
 
+		if newEv.startTimestamp != ev.startTimestamp {
+			// Adjust the offset of selectors based on the new
+			// start time of the evaluator since the calculation
+			// of the offset with @ happens w.r.t. the start time.
+			setOffsetForAtModifier(newEv.startTimestamp, e.Expr)
+		}
+
 		res, ws := newEv.eval(e.Expr)
 		ev.currentSamples = newEv.currentSamples
+		if newEv.startTimestamp != ev.startTimestamp {
+			// Adjust back the offset of selectors based on the parent evaluator.
+			setOffsetForAtModifier(ev.startTimestamp, e.Expr)
+		}
+		return res, ws
+	case *parser.StepInvariantExpr:
+		switch ce := e.Expr.(type) {
+		case *parser.StringLiteral, *parser.NumberLiteral:
+			return ev.eval(ce)
+		}
+
+		if e.Result != nil {
+			return e.Result, nil
```

Comment:
I ran some manual tests and I figured that we don't need to store the result of step invariant node execution at all! Because this will be run only once to get the result for all the steps. With that `matrixIterSlice` function will take care of decrementing the currentSamples.

I will keep this comment open to add a test case where this matrix is dropped mid-execution.